Inline Case Search Parent Select

Jtang-1 opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Adds support for the following workflows:
* Parent select
  * Inline search module can be the target of ‘parent select’ with a 'parent' relationship
  * Inline search module can use parent select with relationship = ‘parent’
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

Was partly implemented here https://github.com/dimagi/commcare-hq/pull/33635
[Tech Spec](https://docs.google.com/document/d/1hp6b3ilLFDF6YYwLLgcr18v4SNR089lOBlLrYDZMOBw/edit)
Jira Ticket: [USH-3578](https://dimagi-dev.atlassian.net/browse/USH-3578)

The reason the workflows did not previously work are partly explained in this [PR](https://github.com/dimagi/commcare-hq/pull/31752).
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
[Inline Case Search](https://www.commcarehq.org/hq/flags/edit/inline_case_search/)
## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Locally tested. The blast radius is limited to the inline case search FF.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added suite tests to verify suite is as expected for the following configurations:
- Inline search module with 'parent select' relationship is 'parent'
- Inline search module as child submodule with 'parent select''parent' relationship. Parent module is also inline search.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
This PR will affect new suite builds for apps that use inline case search. Apps built with this change will not be affected by this PR reversion. But built Apps will need to be rebuilt to revert this PR changes.

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-3578]: https://dimagi-dev.atlassian.net/browse/USH-3578?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Jtang-1 committed:
allows app build with both parent-select and inline case search on the parent/child module

Jtang-1 committed:
feat: exclude cases already in local db from being claimed

Jtang-1 committed:
feat: claims parent cases if parent select is configured and active for the module

Jtang-1 committed:
feat: adds validation that module and its parent select module has unique instance names

Jtang-1 committed:
test: update for excluding claim if case already exists in local casedb

Jtang-1 committed:
test: now supports inline module as target of parent select and inline search using parent select with 'parent' relationship

Jtang-1 committed:
Merge branch 'master' into jt/inline_case_search_parent_select

Jtang-1 committed:
test: claim cases resulting from parent-select's datum

Jtang-1 committed:
test: checks suite file for module with inline case search with 'parent' relationship parent select

Jtang-1 committed:
Merge branch 'master' into jt/inline_case_search_parent_select

Jtang-1 committed:
test: update tests so parent modules also configured with inline search with a 'parent' parent select relationship

Jtang-1 committed:
test: update comment + use constant

Jtang-1 committed:
test: add custom instance names - required when parent and child both use inline search

Jtang-1 committed:
test: update for excluding claim if case already exists in local casedb

Jtang-1 commented:
@snopoke You had this [comment](https://github.com/dimagi/commcare-hq/pull/33617#discussion_r1357944124) in the previous PR that was closed. Wanted to make sure that it was addressed.

Jtang-1 committed:
lint

snopoke commented on corehq/apps/app_manager/tests/test_suite_inline_search.py:
```diff
@@ -380,7 +376,10 @@ def test_inline_search_with_parent_select(self):
             <form>xmlns1.0</form>
             <post url="http://localhost:8000/a/test_domain/phone/claim-case/"
                 relevant="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id]) = 0">
-              <data key="case_id" ref="instance('commcaresession')/session/data/case_id"/>
+              <data exclude="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id]) != 0"
+              key="case_id" ref="instance('commcaresession')/session/data/case_id"/>
+              <data exclude="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_case]) != 0"
+              key="case_id" ref="instance('commcaresession')/session/data/case_id_case"/>
```
If we're including the `exclude` attribute on each case ID data element then we must remove it from the `post` element. Alternatively `post.relevant` must be updated to do both relevancy checks.

Jtang-1 replied:
> If we're including the exclude attribute on each case ID data element then we must remove it from the post element.

Are you saying the `exclude` attribute needs to be removed from each `data` element  in the `post` element?

> both relevancy checks

do you mean `post.relevant` should be:
`relevant=".../data/case_id]) = 0 and ...//data/case_id_case]) = 0` for each case ID data element?

I'm a bit confused, maybe the answer to my question [here](https://github.com/dimagi/commcare-hq/pull/33738#discussion_r1416202776) will also clarify this

snopoke replied:
As it stands now, the `post.relevant` expression only checks if the case from this module doesn't exists. But if the case from the parent module doesn't exist and the case from this module does then the request won't fire.

I think the ideal for this use case would be remove the `relevant` on the post and only have the `data.exclude` expressions. We'd then need to update the Formplayer logic to only do the post request if there is at least one data element that hasn't been excluded.

@shubham1g5 what do you think?

shubham1g5 replied:
We should be able to set `post.relevant` to `case_id != ''` to achieve this without FP changes.

Jtang-1 replied:
@snopoke What do you think about Shubham's suggestion?

snopoke replied:
that seems great

Jtang-1 replied:
Thanks and just noting a couple findings.

1. From Shubham's recommendation - I checked that [evaluatedParams](https://github.com/dimagi/commcare-core/blob/fe5114b44ebe4128205a903be9d69d3d066baf23/src/main/java/org/commcare/suite/model/PostRequest.java#L73C1-L74C1) contains the `case_ids` for all the data_elements(if not excluded)
2. There was a bug where `$case_id!=''` was evaluating as `true` even if all data elements were excluded. I opened this [PR](https://github.com/dimagi/commcare-core/pull/1390) to address it
3. `entry` elements contain computed datums that contain a new uuid case_id
 `datum id="case_id_new_teacher_0" function="uuid()"/>`.
  As is, a data element is created for `case_id_new_teacher_0`. This would cause `$case_id!=''` to always evaluated to `True`, which causes `error unable to recognize next screen: org.commcare.formplayer.screens.FormplayerSyncScreen@d3d077a:`
So my next step is to omit the respective data element.

Please let me know if any of that doesn't seem correct

snopoke commented on corehq/apps/app_manager/tests/test_suite_inline_search.py:
```diff
@@ -419,6 +418,67 @@ def test_inline_search_with_parent_select(self):
+            <session>
+              <datum id="parent_id" nodeset="instance('casedb')/casedb/case[@case_type='case'][@status='open']"
+                value="./@case_id" detail-select="m2_case_short"/>
+              <query url="http://localhost:8000/a/test_domain/phone/search/123/"
+                storage-instance="{RESULTS_INSTANCE_INLINE}" template="case" default_search="false" dynamic_search="false">
+                <title>
+                  <text>
+                      <locale id="case_search.m0.inputs"/>
+                  </text>
+                </title>
+                <data key="case_type" ref="'case'"/>
+                <prompt key="name">
+                  <display>
+                    <text>
+                      <locale id="search_property.m0.name"/>
+                    </text>
+                  </display>
+                </prompt>
+              </query>
```
This should also have an `ancestor-exists` query (or whatever filter is needed to match the 'parent select' config) otherwise the results that are returned may may not be child cases of the parent case.

Jtang-1 replied:
That makes sense. Is this what we need? [cbc5d4c](https://github.com/dimagi/commcare-hq/pull/33738/commits/cbc5d4c5b75f077f069ea787837b081c68a537a6)

Jtang-1 replied:
Wait actually I don't really get it. Within the session block, there's the datum 
```
<datum id="case_id"
 nodeset="instance('{RESULTS_INSTANCE_INLINE}')/results/case[@case_type='case'][@status='open'][active = 'yes'][not(commcare_is_related_case=true())][index/parent=instance('commcaresession')/session/data/parent_id]"
 ```
Doesn't the `index/parent` portion filter such that returned cases will be childs of the parent case? And this is referenced in the post data 
```
<data exclude="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id]) != 0"
 key="case_id" ref="instance('commcaresession')/session/data/case_id"/>
```

snopoke replied:
The `index/parent` filtering only happens after the search results are returned from ES so if we include a bunch of cases that aren't children of the selected parent then we end up filtering them out in FP. In an extreme case, all the search results returned may be filtered out leaving an empty set of results even though there could be child cases in ES.

Jtang-1 replied:
oh i see, that makes sense. I didn't know that this was where the ES filtering vs post ES filtering was defined. Thanks for explaining

snopoke commented on corehq/apps/app_manager/tests/test_suite_inline_search.py:
```diff
@@ -550,7 +616,12 @@ def test_child_module_with_inline_search_entry(self):
             <post url="http://localhost:8000/a/test_domain/phone/claim-case/"
                 relevant="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_case]) = 0">
-              <data key="case_id" ref="instance('commcaresession')/session/data/case_id_case"/>
+              <data exclude="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_case]) != 0"
+                key="case_id" ref="instance('commcaresession')/session/data/case_id_case"/>
+              <data exclude="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/parent_id]) != 0"
+                key="case_id" ref="instance('commcaresession')/session/data/parent_id"/>
+              <data exclude="count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id]) != 0"
+                key="case_id" ref="instance('commcaresession')/session/data/case_id"/>
```
How come there are 3 data elements here? Looking at the session element there should only be `case_id_case` and `case_id`. Where does `parent_id` come from?

Jtang-1 replied:
`get_datum_meta_module` correctly returns the chain of parent select `session_var`. In this test, if submodule isn't being used, the session datums would have the IDs `parent_id` and `case_id`. By configuring the child parent select module as the submodule, `parent_id` changes to `case_id` and `case_id`  changes to `case_id_case`.

Still looking into where/why this happens. I'm suspecting it might be part of the post process, so I might have to change figuring out the parent datum ID to occur in the post process step as you suggested in a previous [comment](https://github.com/dimagi/commcare-hq/pull/33617#discussion_r1357944124).

Jtang-1 replied:
Ah, `get_case_datums_basic_module` is what creates the final case datums  [05683ae](https://github.com/dimagi/commcare-hq/pull/33738/commits/05683aecda28234bd997f8682bfd1ebdd660e1c2)

Jtang-1 commented on corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:
```diff
@@ -153,6 +153,8 @@ def build_case_id_query_data(self):
         else:
             data.ref = QuerySessionXPath(self.case_session_var).instance()
+            if not self.exclude_relevant:
+                data.exclude = CaseIDXPath(data.ref).case().count().neq(0)
```
I thought this (not claiming a case if it already exists in the local case DB) is always relevant? 

What would be the conditional check for when it's needed? Is it for only inline case search modules (so adding a `module_uses_inline_search(self.module)` check)?

Jtang-1 replied:
oh, would it be needed only for <data> elements inside <query> datum (so not post?)

Jtang-1 replied:
Ah got it. That makes sense. Implemented [4d44e5b](https://github.com/dimagi/commcare-hq/pull/33738/commits/4d44e5b64ddf0143ce86442ee29cff6217bf9d87)

Note on the implementation: I think a more general way would have been to check within `build_case_id_query_data` if there will be multiple claims that will be claimed. But the other cases aren't added until later in `add_post_to_entry`. So I couldn't think of a way other than doing this very specific conditional check.

Jtang-1 committed:
feat+test: filter for only cases that are children of "parent select" target module's case type

otherwise the results that are returned may not be child cases of the parent case.

Jtang-1 committed:
test: filters for child cases that have parents that are the target "parent-select"'s case type

the test differentiates the case type between the child and parent to check that the correct
case type is uesd for the <data> ref

Jtang-1 commented on corehq/apps/app_manager/tests/test_suite_inline_search.py:
```diff
@@ -559,12 +630,28 @@ def test_child_module_with_inline_search_entry(self):
+            <instance id="results:parent_instance" src="jr://instance/remote/results:parent_instance"/>
             <session>
-              <datum id="case_id" nodeset="instance('casedb')/casedb/case[@case_type='case'][@status='open']"
+              <query default_search="false" dynamic_search="false" storage-instance="{self.m0.search_config.get_instance_name()}" template="case" url="http://localhost:8000/a/test_domain/phone/search/123/">
+                <title>
+                  <text>
+                    <locale id="case_search.m0.inputs"/>
+                  </text>
+                </title>
+                <data key="case_type" ref="'case'"/>
+                <prompt key="name">
+                  <display>
+                    <text>
+                      <locale id="search_property.m0.name"/>
+                    </text>
+                  </display>
+                </prompt>
+              </query>
+              <datum id="case_id" nodeset="instance('{self.m0.search_config.get_instance_name()}')/results/case[@case_type='case'][@status='open'][not(commcare_is_related_case=true())]"
```
Ya the `case_id` datum is coming from the parent module due to `parent-select` (parent relationship). With "parent-select", isn't it supposed to present the search screen twice? The workflow would be :
Select Submodule Menu -> Case search parent case -> case search cases that are children of selected parent case -> open form

>  we should be able to get the case ID from the commcaresession instance instead of re-querying for it.

Are you talking about the case ID of the parent case we're interested in?  The `query` element populates the instance that the`<datum id="case_id"...>` nodeset references right? I'm not following how `case_id` would be populated in the `commcaresession` instance without the `query` element

shubham1g5 replied:
> Because this is a child module, the user would already have selected the parent case by the time we get to this point in the navigation so the case ID would already be in the session.

but the child case id still needs to be selected right ? 


>  Looking at how normal child modules work I see they still have the full nodeset expression. I

But they should filter on parent case selected I guess ? 


> I guess we'll just have to test this and see if it presents an extra search screen or not.

Would the child cases be in the search results instance though ? If yes,  I guess the search screen is not really useful and we can just write the nodeset for child datum so that it filters the results instance with parent case selected to only show the related cases.

Jtang-1 replied:
@snopoke Could you clarify some more, I'm still unsure how to proceed from here.

I looked into it, and this query is added because the parent module `auto_launch=True`, meaning it's always going to case search first. If i set `auto_launch=False`, then that `query` element is no longer present. And the `datum` you highlighted becomes this:
`<datum detail-select="m0_case_short" id="case_id" nodeset="instance('casedb')/casedb/case[@case_type='case'][@status='open']" value="./@case_id"/>`
which gets the parent case from `casedb`.

Otherwise, by extra screen, do you mean a 3rd screen? When I test it on Web Apps, I see two search screens which is what I expect.
Select Parent Module -> Select Submodule Menu -> Case search parent case (Search Screen1)-> case search for cases that are children of selected parent case (SearchScreen2)-> select form
SearchScreen1 collects the data for datum `case_id` and Search Screen2 collects the data for datum `case_id_case`

snopoke replied:
OK, that seems fine then. I think what Shubham was suggesting is a possible optimization but I'm not sure it would always work.

snopoke commented on corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:
```diff
@@ -295,6 +299,17 @@ def _remote_request_query_datums(self):
+            datums.append(
+                QueryData(
+                    key=CASE_SEARCH_XPATH_QUERY_KEY,
+                    ref=f"ancestor-exists(parent, @case_type={parent_case_type})"
```
add string quotes:
```suggestion
                    ref=f"ancestor-exists(parent, @case_type='{parent_case_type}')"
```

Jtang-1 replied:
[ea5b9f3](https://github.com/dimagi/commcare-hq/pull/33738/commits/ea5b9f31cbaa103c4bbb5dc74755eb6b45367bbf) and [f8b91a7](https://github.com/dimagi/commcare-hq/pull/33738/commits/f8b91a76d9b24e2f1bf414cba22b329dc9761139)

Jtang-1 committed:
bug: gets the final case datums after processing. get_datum_meta_module contains unprocessed datums which ids are changed later in the entry creation

Jtang-1 committed:
refactor: gets case_datum_id only if it's needed

Jtang-1 committed:
feat: further restrict conditions for adding all case datums <data> blocks to post block

Jtang-1 committed:
bug: case search query string requires quote

Co-authored-by: Simon Kelly <skelly@dimagi.com>

Jtang-1 committed:
test: corrects case search query string

Jtang-1 committed:
bug: ref string requires to be enclosed by quotes

Jtang-1 committed:
makes multi_select_relevant into class method

Jtang-1 committed:
test: corrects case search query to be enclosed in quotes

Jtang-1 committed:
relevant check needs to account for all data elements within post.

Jtang-1 committed:
bug: datums for new cases should not in post.data as we do not need to claim these new cases

Jtang-1 committed:
only have "exclude" attribute to  data element if only claiming a single case

there is already a "relevant" attribute on the "post" element that does the check when we're
claimining a single case. Having the "exclude" on the data element with the same xpath expression
means we end up checking twice for the same case.

Jtang-1 commented:
@snopoke Thank you for the review/comments. I've addressed and replied to them all with respective commits. Please let me know if there's anything I missed

snopoke commented on corehq/apps/app_manager/suite_xml/post_process/remote_requests.py:
```diff
@@ -306,7 +306,7 @@ def _remote_request_query_datums(self):
                 QueryData(
                     key=CASE_SEARCH_XPATH_QUERY_KEY,
-                    ref=f"ancestor-exists(parent, @case_type='{parent_case_type}')"
+                    ref=f"\"ancestor-exists(parent, @case_type='{parent_case_type}')\""
```
oof

snopoke commented on corehq/apps/app_manager/suite_xml/sections/entries.py:
```diff
@@ -266,6 +267,7 @@ def add_post_to_entry(self, form, module, e):
                 data.ref = QuerySessionXPath(case_datum_id).instance()
                 data.exclude = CaseIDXPath(data.ref).case().count().neq(0)
                 e.post.data.append(data)
+            e.post.relevant = CaseClaimXpath.multi_select_relevant()
```
maybe rename this to `mutli_case_relevant` since it now applies to more than just multi-select.

Jtang-1 replied:
good call [ba15267](https://github.com/dimagi/commcare-hq/pull/33738/commits/ba1526721640941be0ea8ad5d29ebb47ebf3a83e)

snopoke reviewed (approved):
Thanks for all the fixes. This is looking good to me.

Is there a QA plan? I don't see anything noted in the PR description

Jtang-1 committed:
rename variable for clarity

Jtang-1 commented:
I wasn't planning on QA but it is worth doing. The changes for the most part are gated behind specific configuration checks so it wouldn't affect anything existing, but QA would be good to check this new workflow.

Given that and the urgency for delivery to be able to start using this, I'm thinking of having QA test it out after deploy.

Also, thought of this while typing. I think the validation should require version 2.54 since this configuration requires the corresponding FP changes

shubham1g5 commented:
> I think the validation should require version 2.54 since this configuration requires the corresponding FP changes

These workflows doesn't really work on Mobile so it's ok to skip this validation (as I think there are no breaking model changes for mobile)

Jtang-1 commented:
I asked Shubham to approve for the last commit [ba15267](https://github.com/dimagi/commcare-hq/pull/33738/commits/ba1526721640941be0ea8ad5d29ebb47ebf3a83e) so I can merge
